app/frontend/wallet/shelley/transaction/computeTxPlan.ts

```diff
@@ -0,0 +1,268 @@
+    return {
+      success: false,
+      minimalLovelaceAmount: additionalLovelaceAmount,
+      estimatedFee: feeWithoutChange,
+      deposit,
+      error: {code: UnexpectedErrorReason.CannotConstructTxPlan},
+    }
+  }
+
+  // No change necessary, perfect fit
+  if (totalOutput + feeWithoutChange === totalInput && tokenDifference.length === 0) {
+    return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: [],
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: feeWithoutChange,
+        baseFee: feeWithoutChange,
+        withdrawals,
+        auxiliaryData,
+      },
+    }
+  }
+
+  // From this point on we are sure its not a perfect fit, so change has to be calculated
+
+  const changeOutputs: TxOutput[] = computeChange(
+    possibleChange.address,
+    tokenDifference,
+    MAX_TX_OUTPUT_SIZE,
+    MAX_OUTPUT_TOKENS
+  )
+
+  const feeWithChange = computeRequiredTxFee(
+    inputs,
+    [...outputs, ...changeOutputs],
+    certificates,
+    withdrawals,
+    auxiliaryData
+  )
+
+  // We cannot fit the change output into the transaction instead, just increase the fee
+  if (tokenDifference.length === 0 && totalOutput + feeWithChange > totalInput) {
+    return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: [],
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: (totalInput - totalOutput) as Lovelace,
+        baseFee: feeWithoutChange,
+        withdrawals,
+        auxiliaryData,
+      },
+    }
+  }
+
+  // We cannot create a change output with minimal ada so we add it to the fee
+  if (
+    tokenDifference.length === 0 &&
+    changeOutputs.length === 1 &&
+    totalInput - totalOutput - feeWithChange < MIN_UTXO_VALUE
+  ) {
+    return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: [],
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: (totalInput - totalOutput) as Lovelace,
+        baseFee: feeWithoutChange,
+        withdrawals,
+        auxiliaryData,
+      },
+    }
+  }
+
+  // From this point on change has to be included in transaction for it to be balanced
+
+  const changeAdditionalLovelaceAmount = changeOutputs.reduce(
+    (acc, {coins}) => acc + coins,
+    0
+  ) as Lovelace
+
+  const lovelaceChange = totalInput - totalOutput - changeAdditionalLovelaceAmount - feeWithChange
+
+  // lovelaceChange is added to the first change output
+  // so it has to be positive to keep its min ada value
+  if (tokenDifference.length > 0 && lovelaceChange < 0) {
+    return {
+      success: false,
+      minimalLovelaceAmount: additionalLovelaceAmount,
+      estimatedFee: feeWithChange,
+      deposit,
+      error: {code: UnexpectedErrorReason.CannotConstructTxPlan},
+    }
+  }
+
+  const firstChangeOutput: TxOutput = {
```

Comment:
isn't there more value in just adding one more ada-only tx output? The ada in token outputs is imho kind of "locked up" so I'd try to "free it up" even at the cost of some fees which the user saves when performing ada-only transactions as the fee is linear

Reply:
though I see - things would get hairy if the amount is less than minimum ada per utxo, though that should be a constant for ada-only utxos

Reply:
Thats a good point regarding the "freeing of ADA only outputs" but other than the const of fees, this would also require the output to be >= 1 ADA, unlike now when it only needs to be positive. What do you think? I